File: cpp-client/deephaven/tests/lastby_test.cc
/*
 * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
 */
#include "tests/third_party/catch.hpp"
#include "tests/test_util.h"

using deephaven::client::TableHandleManager;
using deephaven::client::TableHandle;
using deephaven::dhcore::utility::Streamf;
using deephaven::dhcore::utility::Stringf;

namespace deephaven::client::tests {
TEST_CASE("Last By", "[lastby]") {
  auto tm = TableMakerForTests::Create();
  auto table = tm.Table();

  auto lb = table.Where("ImportDate == `2017-11-01`")
      .Select("Ticker", "Open", "Close")
      .LastBy("Ticker");

  std::vector<std::string> ticker_data = {"XRX", "XYZZY", "IBM", "GME", "AAPL", "ZNGA"};
  std::vector<double> open_data = {50.5, 92.3, 40.1, 681.43, 31.5, 685.3};
  std::vector<double> close_data = {53.8, 88.5, 38.7, 453, 26.7, 544.9};

  INFO(lb.Stream(true));
  CompareTable(
      lb,
      "Ticker", ticker_data,
      "Open", open_data,
      "Close", close_data
  );
}
}  // namespace deephaven::client::tests

File: csharp/client/Dh_NetClientTests/LastByTest.cs
//
// Copyright (c) 2016-2025 Deephaven Data Labs and Patent Pending
//
using Deephaven.Dh_NetClient;
using Xunit.Abstractions;

namespace Deephaven.Dh_NetClientTests;

public class LastByTest(ITestOutputHelper output) {
  private readonly ITestOutputHelper _output = output;

  [Fact]
  public void TestLastBy() {
    using var ctx = CommonContextForTests.Create(new ClientOptions());
    var testTable = ctx.TestTable;

    using var lb = testTable.Where("ImportDate == `2017-11-01`")
      .Select("Ticker", "Open", "Close")
      .LastBy("Ticker");

    var tickerData = new[] {
      "XRX", "XYZZY", "IBM", "GME", "AAPL", "ZNGA"
    };

    var openData = new[] { 50.5, 92.3, 40.1, 681.43, 31.5, 685.3 };
    var closeData = new[] { 53.8, 88.5, 38.7, 453, 26.7, 544.9 };
    var expected = new TableMaker();
    expected.AddColumn("Ticker", tickerData);
    expected.AddColumn("Open", openData);
    expected.AddColumn("Close", closeData);

    TableComparer.AssertSame(expected, lb);
  }
}
